connectors/kafka/src/main/java/forklift/producers/KafkaForkliftProducer.java

```diff
@@ -0,0 +1,257 @@
+                                                   "\"doc\":\"Properties added to support forklift interfaces. Format is key,value entries delimited with new lines\"}");
+        ObjectNode schemaNode = (ObjectNode)mapper.readTree(originalJson);
+        ArrayNode fieldsNode = (ArrayNode)schemaNode.get("fields");
+        fieldsNode.add(propertiesField);
+        schemaNode.set("fields", fieldsNode);
+        Schema.Parser parser = new Schema.Parser();
+        return parser.parse(mapper.writeValueAsString(schemaNode));
+    }
+
+    private GenericRecord addForkliftPropertiesToAvroObject(SpecificRecord message) throws IOException {
+        //Write message to json
+        ByteArrayOutputStream outputStream = new ByteArrayOutputStream();
+        Encoder encoder = EncoderFactory.get().jsonEncoder(message.getSchema(), outputStream);
+        DatumWriter<SpecificRecord> writer = new SpecificDatumWriter<>(message.getSchema());
+        writer.write(message, encoder);
+        encoder.flush();
+        String json = new String(outputStream.toByteArray(), Charset.forName("UTF-8"));
+
+        //modify schema to include forklift properties
+        Schema modifiedSchema = avroSchemaCache.get(message.getClass());
+        if (modifiedSchema == null) {
+            modifiedSchema = addForkliftPropertiesToSchema(message.getSchema());
+            avroSchemaCache.put(message.getClass(), modifiedSchema);
+        }
+
+        //add forklift properties to json
+        ObjectNode messageNode = (ObjectNode)mapper.readTree(json);
+        messageNode.put(SCHEMA_FIELD_NAME_PROPERTIES, this.formatMap(this.properties));
+
+        //read modified json to avro object with modified schema
+        InputStream input = new ByteArrayInputStream(messageNode.toString().getBytes(Charset.forName("UTF-8")));
+        DataInputStream din = new DataInputStream(input);
+        Decoder decoder = DecoderFactory.get().jsonDecoder(modifiedSchema, din);
+        DatumReader<GenericRecord> reader = new GenericDatumReader<>(modifiedSchema);
+        return reader.read(null, decoder);
+    }
+
+    private String sendAvroMessage(SpecificRecord message) throws ProducerException {
+        try {
+            GenericRecord avroRecord = addForkliftPropertiesToAvroObject(message);
+            ProducerRecord record = new ProducerRecord<String, GenericRecord>(topic, null, avroRecord);
+            try {
+                RecordMetadata result = (RecordMetadata)kafkaProducer.send(record).get();
+                return result.topic() + "-" + result.partition() + "-" + result.offset();
+            } catch (InterruptedException e) {
+                Thread.currentThread().interrupt();
+                throw new ProducerException("Error sending Kafka Message", e);
+            } catch (ExecutionException e) {
+                throw new ProducerException("Error creating Kafka Message", e);
```

Comment:
This exception message doesn't make sense; the message is completely created before the corresponding `try` block.